### PR TITLE
Meta: Replace `__filebasename` with `import.meta.url`

### PR DESCRIPTION
**Pull request by cheap-glitch**
https://github.com/refined-github/refined-github/issues/4008#issuecomment-962594700

- follows https://github.com/refined-github/refined-github/issues/4320

This PR:
 * changes `features.add()` and `features.addCssFeature()` to take the file URL as their first argument
 * adds `features.getFeatureID()`

### CODE REVIEW

File: source/globals.d.ts [hunk 1]
```diff
@@ -2,15 +2,13 @@
 
 type AnyObject = Record<string, any>;
 
-type FeatureID = 'use the __filebasename variable';
+type FeatureID = string;
```
Comment (cheap-glitch) [inline source/globals.d.ts:5]: We could use an opaque type here, to enforce better semantics.

File: source/features/index.tsx [hunk 2]
```diff
@@ -211,12 +211,15 @@ function enforceDefaults(
 	}
 }
 
+const getFeatureID = (url: string): FeatureID => url.split('/').pop()!.split('.', 1)[0];
```
Comment (yakov116) [inline source/features/index.tsx:214]: I think regex would be more readable. WDYT?

Comment (cheap-glitch) [reply source/features/index.tsx:214]: > I think regex would be more readable. WDYT?

I don't think so. This seems simple enough to me

File: source/features/index.tsx [hunk 3]
```diff
@@ -156,7 +156,7 @@ const setupPageLoad = async (id: FeatureID, config: InternalRunConfig): Promise<
 		try {
 			// Features can return `false` when they decide not to run on the current page
 			// Also the condition avoids logging the fake feature added for `has-rgh`
-			if (await init() !== false && !id?.startsWith('rgh') && id !== __filebasename) {
+			if (await init() !== false && !id?.startsWith('rgh') && id !== getFeatureID(import.meta.url)) {
```
Comment (fregante) [inline source/features/index.tsx:159]: 1/2
```suggestion
			if (await init() !== false && !id?.startsWith('rgh')) {
```

File: source/features/index.tsx [hunk 4]
```diff
@@ -283,7 +287,7 @@ This means that the old features will still be on the page and don't need to re-
 
 This marks each as "processed"
 */
-void add(__filebasename, {
+void add(import.meta.url, {
```
Comment (fregante) [inline source/features/index.tsx:290]: 2/2
```suggestion
void add('rgh-deduplicator' as FeatureID, {
```

File: source/features/index.tsx [hunk 5]
```diff
@@ -211,12 +211,15 @@ function enforceDefaults(
...
+
 /** Register a new feature */
-const add = async (id: FeatureID, ...loaders: FeatureLoader[]): Promise<void> => {
+const add = async (url: string, ...loaders: FeatureLoader[]): Promise<void> => {
+	const id = getFeatureID(url);
 	/* Feature filtering and running */
 	const options = await globalReady;
 	// Skip disabled features, unless the "feature" is the fake feature in this file
-	if (!options[`feature:${id}`] && id as string !== __filebasename) {
+	if (!options[`feature:${id}`] && id !== getFeatureID(import.meta.url)) {
```
Comment (fregante) [inline source/features/index.tsx:222]: 3/2

We can probably also use the `rgh-` check here 
```suggestion
	if (!options[`feature:${id}`] && !id.startsWith('rgh')) {
```

File: source/features/mark-private-orgs.tsx [hunk 6]
```diff
@@ -16,7 +16,7 @@ const getPublicOrganizationsNames = cache.function(async (username: string): Pro
 	return response.map((organization: AnyObject) => organization.login);
 }, {
 	maxAge: {days: 10},
-	cacheKey: ([username]) => __filebasename + ':' + username,
+	cacheKey: ([username]) => features.getFeatureID(import.meta.url) + ':' + username,
```
Comment (fregante) [inline source/features/mark-private-orgs.tsx:19]:
```suggestion
	cacheKey: ([username]) => 'public-organizations:' + username,
```

File: source/features/repo-age.tsx [hunk 7]
```diff
@@ -92,7 +92,7 @@ const getFirstCommit = cache.function(async (): Promise<[committedDate: string,
 
 	return getRepoAge(commitSha, commitsCount);
 }, {
-	cacheKey: () => __filebasename + ':' + getRepo()!.nameWithOwner,
+	cacheKey: () => features.getFeatureID(import.meta.url) + ':' + getRepo()!.nameWithOwner,
```
Comment (fregante) [inline source/features/repo-age.tsx:95]:
```suggestion
	cacheKey: () => 'first-commit:' + getRepo()!.nameWithOwner,
```

File: source/features/submit-review-as-single-comment.tsx [hunk 8]
```diff
@@ -56,7 +56,7 @@ async function handleSubmitSingle(event: delegate.Event): Promise<void> {
 	const commentText = select('textarea[name="pull_request_review_comment[body]"]', commentContainer)!.value;
 	if (!commentText) {
 		alert('Error: Comment not found and not submitted. More info in the console.');
-		features.log.error(__filebasename, 'Comment not found');
+		features.log.error(features.getFeatureID(import.meta.url), 'Comment not found');
```
Comment (fregante) [inline source/features/submit-review-as-single-comment.tsx:59]: `log.error` should also be updated to accept the raw url

```suggestion
		features.log.error(import.meta.url, 'Comment not found');
```

File: source/features/user-local-time.tsx [hunk 9]
```diff
@@ -67,7 +67,7 @@ const getLastCommitDate = cache.function(async (login: string): Promise<string |
 }, {
 	maxAge: {days: 10},
 	staleWhileRevalidate: {days: 20},
-	cacheKey: ([login]) => __filebasename + ':' + login,
+	cacheKey: ([login]) => features.getFeatureID(import.meta.url) + ':' + login,
```
Comment (fregante) [inline source/features/user-local-time.tsx:70]:
```suggestion
	cacheKey: ([login]) => 'last-commit:' + login,
```

File: source/globals.d.ts (hunk 1 above)
Comment (fregante) [reply source/globals.d.ts:5]: That's what this was. It should be undone

```suggestion
type FeatureID = 'use the __filebasename variable';
```

**Review by fregante — CHANGES_REQUESTED** [review]

File: source/features/bugs-tab.tsx [hunk 10]
```diff
@@ -63,7 +63,7 @@ const countBugs = cache.function(async (): Promise<number> => {
 }, {
 	maxAge: {minutes: 30},
 	staleWhileRevalidate: {days: 4},
-	cacheKey: (): string => __filebasename + ':' + getRepo()!.nameWithOwner,
+	cacheKey: (): string => features.getFeatureID(import.meta.url) + ':' + getRepo()!.nameWithOwner,
```
Comment (fregante) [inline source/features/bugs-tab.tsx:66]:
```suggestion
	cacheKey: (): string => 'bugs:' + getRepo()!.nameWithOwner,
```

File: source/features/list-prs-for-file.tsx [hunk 11]
```diff
@@ -103,7 +103,7 @@ const getPrsByFile = cache.function(async (): Promise<Record<string, number[]>>
 }, {
 	maxAge: {hours: 2},
 	staleWhileRevalidate: {days: 9},
-	cacheKey: () => __filebasename + ':' + getRepo()!.nameWithOwner,
+	cacheKey: () => features.getFeatureID(import.meta.url) + ':' + getRepo()!.nameWithOwner,
```
Comment (fregante) [inline source/features/list-prs-for-file.tsx:106]:
```suggestion
	cacheKey: () => 'files-with-prs:' + getRepo()!.nameWithOwner,
```

File: source/features/pr-filters.tsx [hunk 12]
```diff
@@ -82,7 +82,7 @@ const hasChecks = cache.function(async (): Promise<boolean> => {
 	return repository.head.history.nodes.some((commit: AnyObject) => commit.statusCheckRollup);
 }, {
 	maxAge: {days: 3},
-	cacheKey: () => __filebasename + ':' + getRepo()!.nameWithOwner,
+	cacheKey: () => features.getFeatureID(import.meta.url) + ':' + getRepo()!.nameWithOwner,
```
Comment (fregante) [inline source/features/pr-filters.tsx:85]:
```suggestion
	cacheKey: () => 'has-checks:' + getRepo()!.nameWithOwner,
```

**Review by fregante — APPROVED** [review]

**Pull request merged**